pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/segment/OfflineSegmentAssignment.java

```diff
@@ -0,0 +1,260 @@
+ *   </li>
+ *   <li>
+ *     Replica-group based assignment (more than 1 replicas in instance partitions):
+ *     <p>Among replicas, always mirror the assignment (pick the same index of the instance).
+ *     <p>Within each partition, assign the segment to the servers with the least segments already assigned. In case of
+ *     a tie, assign to the server with the smallest index in the list. Do this for one replica and mirror the
+ *     assignment to other replicas.
+ *     <p>To rebalance a table, within each partition, first calculate the number of segments on each server, loop over
+ *     all the segments and keep the assignment if number of segments for the server has not been reached and track the
+ *     not assigned segments, then assign the left-over segments to the servers with the least segments, or the smallest
+ *     index if there is a tie. Repeat the process for all the partitions in one replica, and mirror the assignment to
+ *     other replicas. With this greedy algorithm, the result is deterministic and with minimum segment moves.
+ *   </li>
+ * </ul>
+ */
+public class OfflineSegmentAssignment implements SegmentAssignment {
+  private static final Logger LOGGER = LoggerFactory.getLogger(OfflineSegmentAssignment.class);
+
+  private HelixManager _helixManager;
+  private String _offlineTableName;
+  private int _replication;
+  private String _partitionColumn;
+
+  @Override
+  public void init(HelixManager helixManager, TableConfig tableConfig) {
+    _helixManager = helixManager;
+    _offlineTableName = tableConfig.getTableName();
+    _replication = tableConfig.getValidationConfig().getReplicationNumber();
+    ReplicaGroupStrategyConfig strategyConfig = tableConfig.getValidationConfig().getReplicaGroupStrategyConfig();
+    _partitionColumn = strategyConfig != null ? strategyConfig.getPartitionColumn() : null;
+
+    if (_partitionColumn == null) {
+      LOGGER.info("Initialized OfflineSegmentAssignment with replication: {} without partition column for table: {} ",
+          _replication, _offlineTableName);
+    } else {
+      LOGGER.info("Initialized OfflineSegmentAssignment with replication: {} and partition column: {} for table: {}",
+          _replication, _partitionColumn, _offlineTableName);
+    }
+  }
+
+  @Override
+  public List<String> assignSegment(String segmentName, Map<String, Map<String, String>> currentAssignment,
+      Map<InstancePartitionsType, InstancePartitions> instancePartitionsMap) {
+    InstancePartitions instancePartitions = instancePartitionsMap.get(InstancePartitionsType.OFFLINE);
+    Preconditions.checkState(instancePartitions != null, "Failed to find OFFLINE instance partitions for table: %s",
+        _offlineTableName);
+    LOGGER.info("Assigning segment: {} with instance partitions: {} for table: {}", segmentName, instancePartitions,
+        _offlineTableName);
+
+    List<String> instancesAssigned;
+    if (instancePartitions.getNumReplicas() == 1) {
```

Comment:
should this method be called getNumReplicaGroups()? It is confusing with the variable _replication (whcih, I suppose can be more than 1 even under this if condition)

Reply:
Will address in another pr